### DISCUSSION TITLE
scene render_to_image() and render_to_depth_image() not the same transform

### DISCUSSION THREAD

**Discussion by jerome-godbout-lychens** [isl-org/Open3D#6996] [Q&A]
Hi,

I try to capture the current image and the depth to perform some 2D detection and bring it back to 3D. When I raycast back to 3D, the unproject work for UV coordinate but I get a wrong depth. I did inspect the depth map that result from the render_to_depth_image() and what stroke me is that the object position is not the same as the actual scene image. The zoom is not even the same, the position either. Can somebody shed some light on this and how it work? 

The strip down example use a cube at the origin and camera with look at, nothing else, the capture is start automatically after 2s. I did try to capture depth before image or image before depth, same results. I did try to add delay between the capture, same results. How is the depth image generated? do I need something special or is the scene look-at not considered somehow??? Both image have the right dimension compare to the window, they just not display the same scene. I also try to invert UV coordinate from the depth map just in case: self._capture_depth[:, [0, 1]] = self._capture_depth[:, [1, 0]] but without much results. Seem like it always use the origine as the top right corder of the image and doesn't care where the camera is point to.

The normalized depth image from render_to_depth_image():
![captured_image_depth](https://github.com/user-attachments/assets/5f74404a-35bf-45dd-94d2-f75375fdd4aa)
 
The actual scene image from render_to_image():
![captured_image](https://github.com/user-attachments/assets/adfe8867-70ec-48bd-80d0-d7a5ab1af3ea)

The demo code used for this (reduce set to minimal example):
[demo_bug.py.txt](https://github.com/user-attachments/files/17247970/demo_bug.py.txt)

Is their any way to get the depth map of the actual rendered scene?

I also try to replace the cube with a pyramide to see if I can better understand what is going on, just seems off for some odd reason:
use this into the def add_cube(self): instead
mesh_box = o3d.geometry.TriangleMesh.create_cone(radius=1.5, height=3.0, resolution=4, split=1, create_uv_map=False)

you can also add the axis into the __init__:
self._scene.scene.show_axes(True)

Any help or tips on what is going on with that would be appreciated thanks,

I did try to use the pyramide and axes and superpose the captured image and depth: and this is the results:
![image](https://github.com/user-attachments/assets/75b31bfa-cf47-4f14-a52c-80f1b011f860)

**Comment by jerome-godbout-lychens** [comment]
I think I have figure out, there is no tranform, it's just the wrong image scale, I'm on a Macbook, seems like the render to image use the pixel density but the depth map doesn't. Resulting into a depth map 1/2 the height and 1/2 the width of the render. So the render to image and the render to depth image have a different way of processing the screen dpi and pixel density. The render to image is right but the render to image depth is not resulting in only a 1/4 of the actual surface.
<img width="1254" alt="image" src="https://github.com/user-attachments/assets/ade8d74f-c012-43fc-8fb9-ecc3a4d1e362">

**Comment by moooeeeep** [comment]
Not sure if I stumbled accross the same problem, or a related one, but I had the impression `open3d.io.write_image`  just doesn't work correctly for (single-channel) depth images. At least I only got garbled images out of it (tested with open3d==0.17.0).

When I switched to write the image with PIL, it worked (copied and adjusted the example code from [here](https://github.com/isl-org/Open3D/issues/4417#issuecomment-993780683)):
```
import numpy as np
import open3d as o3d
import PIL

sphere = o3d.geometry.TriangleMesh.create_sphere(4.0)
sphere.compute_vertex_normals()
cylinder = o3d.geometry.TriangleMesh.create_cylinder(1.0, 4.0, 30, 4)
cylinder.compute_triangle_normals()
cylinder.translate([6, 2, 0.0])

render = o3d.visualization.rendering.OffscreenRenderer(640, 480)
mat = o3d.visualization.rendering.MaterialRecord()
mat.shader = 'defaultLit'

render.scene.add_geometry("sphere1", sphere, mat)
render.scene.add_geometry("cylinder1", cylinder, mat)
render.setup_camera(45, [0, 0, 0], [0, 0, -25.0], [0, 1, 0])

cimg = render.render_to_image()
dimg = render.render_to_depth_image(True)

o3d.io.write_image("cimg.png", cimg, 0)

# this will write a garbled depth image to the file
#o3d.io.write_image("dimg.png", dimg, 0)

# convert to numpy array, apply some normalization for visualization
arr = np.asarray(dimg)
arr[arr==np.inf] = 0.0
arr = (arr / arr.max() * np.iinfo(np.uint16).max).astype(np.uint16)
# write with PIL
PIL.Image.fromarray(arr).save("dimg.png")
```
